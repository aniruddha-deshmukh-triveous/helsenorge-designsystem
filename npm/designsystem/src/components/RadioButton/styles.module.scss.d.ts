export type Styles = {
  'radio-button': string;
  'radio-button--disabled': string;
  'radio-button--invalid': string;
  'radio-button--on-blueberry': string;
  'radio-button--on-dark': string;
  'radio-button-errors': string;
  'radio-button-label': string;
  'radio-button-label--bigform': string;
  'radio-button-label--disabled': string;
  'radio-button-label--invalid': string;
  'radio-button-label--on-blueberry': string;
  'radio-button-label--on-dark': string;
  'radio-button-wrapper': string;
  'radio-button-wrapper--bigform': string;
  'radio-button-wrapper--with-error': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
