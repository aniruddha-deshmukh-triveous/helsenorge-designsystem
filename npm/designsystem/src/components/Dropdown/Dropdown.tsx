import React, { useRef, useState } from 'react';

import classNames from 'classnames';

import {
  AnalyticsId,
  IconSize,
  KeyboardEventKey,
  theme,
  useHover,
  useKeyboardEvent,
  useOutsideEvent,
  useSize,
  useToggle,
  useUuid,
} from '../..';
import Button from '../Button';
import Icon from '../Icons';
import PlusSmall from '../Icons/PlusSmall';

import styles from './styles.module.scss';

export enum DropdownMode {
  onwhite = 'onwhite',
  ongrey = 'ongrey',
  onblueberry = 'onblueberry',
  oncherry = 'oncherry',
}

export interface DropdownProps {
  /** Label for dropdown. Synlig for skjermlesere.  */
  label: string;
  /** Tekst på knappen som åpner dropdownen */
  placeholder: string;
  /** Sets the dropdown content */
  children: React.ReactNode;
  /** Close button text */
  closeText?: string;
  /** No close button */
  noCloseButton?: boolean;
  /** Called when dropdown is open/closed. */
  onToggle?: (isOpen: boolean) => void;
  /** Om dropdown er åpen */
  open?: boolean;
  /** Changes the visuals of the dropdown */
  mode?: keyof typeof DropdownMode;
  /** Makes the background transparent */
  transparent?: boolean;
  /** Makes the background transparent */
  fluid?: boolean;
  /** Makes the dropdown disabled */
  disabled?: boolean;
  /** Sets the data-testid attribute on the dropdown button. */
  testId?: string;
}

const Dropdown: React.FC<DropdownProps> = props => {
  const {
    label,
    placeholder,
    closeText = 'Lukk',
    noCloseButton = false,
    onToggle,
    open = false,
    children,
    mode = DropdownMode.onwhite,
    transparent = false,
    fluid = false,
    testId,
    disabled,
  } = props;
  const dropdownRef = useRef<HTMLDivElement>(null);
  const optionsRef = useRef<HTMLUListElement>(null);
  const { hoverRef: buttonRef, isHovered } = useHover<HTMLButtonElement>();
  const { value: isOpen, toggleValue: toggleIsOpen } = useToggle(!disabled && open, onToggle);
  const inputRefList = useRef(React.Children.map(children, () => React.createRef<HTMLElement>()));
  const [currentIndex, setCurrentIndex] = useState<number>();
  const { width: buttonWidth } = useSize(buttonRef) || {};
  const labelId = useUuid();
  const toggleLabelId = useUuid();
  const optionIdPrefix = useUuid();

  const handleOpen = () => {
    toggleIsOpen();
    optionsRef.current?.focus();
  };

  const handleClose = () => {
    toggleIsOpen();
    buttonRef.current?.focus();
  };

  const handleKeyboardNavigation = (event: KeyboardEvent) => {
    event.preventDefault();

    if (!inputRefList.current) {
      return;
    }

    if (!isOpen) {
      handleOpen();
      return;
    } else if (event.key === KeyboardEventKey.Escape && isOpen) {
      handleClose();
      return;
    }

    const index = inputRefList.current.findIndex(x => x.current === event.target);
    let nextIndex = index;

    if (event.key === KeyboardEventKey.Home) {
      nextIndex = 0;
    } else if (event.key === KeyboardEventKey.End) {
      nextIndex = inputRefList.current.length - 1;
    } else if (event.key === KeyboardEventKey.ArrowDown && index < inputRefList.current.length - 1) {
      nextIndex = index + 1;
    } else if (event.key === KeyboardEventKey.ArrowUp && index > 0) {
      nextIndex = index - 1;
    } else if (event.key === KeyboardEventKey.Enter && index !== -1) {
      nextIndex = index;
    }
    if (nextIndex !== -1) {
      inputRefList.current[nextIndex].current?.focus();
      setCurrentIndex(nextIndex);
    }
  };

  useKeyboardEvent(dropdownRef, handleKeyboardNavigation, [
    KeyboardEventKey.ArrowDown,
    KeyboardEventKey.ArrowUp,
    KeyboardEventKey.End,
    KeyboardEventKey.Enter,
    KeyboardEventKey.Escape,
    KeyboardEventKey.Home,
  ]);

  useOutsideEvent(dropdownRef, () => isOpen && handleClose());

  const toggleClasses = classNames(
    styles.dropdown__toggle,
    !disabled && {
      [styles['dropdown__toggle--on-white']]: mode === DropdownMode.onwhite,
      [styles['dropdown__toggle--on-grey']]: mode === DropdownMode.ongrey,
      [styles['dropdown__toggle--on-blueberry']]: mode === DropdownMode.onblueberry,
      [styles['dropdown__toggle--on-cherry']]: mode === DropdownMode.oncherry,
      [styles['dropdown__toggle--transparent']]: transparent,
      [styles['dropdown__toggle--fluid']]: fluid,
      [styles['dropdown__toggle--open']]: isOpen,
    }
  );

  const contentClasses = classNames(styles.dropdown__content, isOpen && styles['dropdown__content--open']);

  return (
    <div className={styles.dropdown} ref={dropdownRef}>
      <span id={labelId} className={styles.dropdown__label}>
        {label}
      </span>
      <button
        type="button"
        onClick={() => !isOpen && handleOpen()}
        className={toggleClasses}
        ref={buttonRef}
        data-testid={testId}
        data-analyticsid={AnalyticsId.Dropdown}
        disabled={disabled}
        aria-labelledby={toggleLabelId}
        aria-haspopup="listbox"
        aria-expanded={isOpen}
      >
        <span id={toggleLabelId} className={styles.dropdown__toggle__label}>
          {placeholder}
        </span>
        <Icon
          color={disabled ? theme.palette.neutral500 : theme.palette.blueberry600}
          svgIcon={PlusSmall}
          className={styles.dropdown__icon}
          isHovered={isHovered}
          size={IconSize.XSmall}
        />
      </button>
      <div className={contentClasses} style={{ width: fluid ? '100%' : `${buttonWidth}px` }}>
        <ul
          className={styles.dropdown__options}
          role="listbox"
          aria-labelledby={labelId}
          tabIndex={-1}
          aria-activedescendant={typeof currentIndex !== 'undefined' ? `${optionIdPrefix}-${currentIndex}` : undefined}
          ref={optionsRef}
        >
          {React.Children.map(children, (child, index) => (
            <li className={styles.dropdown__input} role="option" id={`${optionIdPrefix}-${index}`}>
              {React.cloneElement(child as React.ReactElement, { ref: inputRefList.current?.[index] })}
            </li>
          ))}
        </ul>
        {!noCloseButton && (
          <div className={styles.dropdown__close}>
            <Button onClick={handleClose} fluid aria-expanded={isOpen}>
              {closeText}
            </Button>
          </div>
        )}
      </div>
    </div>
  );
};

export default Dropdown;
