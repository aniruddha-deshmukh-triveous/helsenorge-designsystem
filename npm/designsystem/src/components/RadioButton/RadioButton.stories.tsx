import React from 'react';

import { ComponentStory, ComponentMeta } from '@storybook/react';

import RadioButton from './RadioButton';
import { FormMode, FormSize } from '../../constants';
import { getColor } from '../../theme/currys';
import FormGroup from '../FormGroup';
import GridExample from '../GridExample';
import Label from '../Label';

export default {
  title: '@helsenorge∕designsystem-react/Components/RadioButton',
  component: RadioButton,
  parameters: {
    docs: {
      description: {
        component:
          'RadioButton lar brukeren velge et av flere valg i en liste. RadioButton kan brukes frittstående, som en del av en FormGroup eller direkte i et Validation komponent.',
      },
    },
  },
  argTypes: {
    label: {
      control: 'text',
      defaultValue: 'RadioButton label',
    },
    defaultChecked: {
      control: 'boolean',
      defaultValue: false,
    },
    disabled: {
      control: 'boolean',
      defaultValue: false,
    },
    mode: {
      control: 'select',
      options: FormMode,
      defaultValue: FormMode.onwhite,
    },
    size: {
      control: 'select',
      options: FormSize,
      defaultValue: FormSize.medium,
    },
    name: {
      control: 'text',
      defaultValue: 'radio',
    },
    value: {
      control: 'text',
      defaultValue: '',
    },
    required: {
      control: 'boolean',
      defaultValue: false,
    },
  },
} as ComponentMeta<typeof RadioButton>;

export const Default: ComponentStory<typeof RadioButton> = args => (
  <GridExample>
    <RadioButton {...args} label={<Label labelTexts={[{ text: 'Radio onwhite' }]} />} mode={'onwhite'} />
    <RadioButton {...args} label={<Label labelTexts={[{ text: 'ongrey' }]} />} mode={'ongrey'} />
    <RadioButton {...args} label={<Label labelTexts={[{ text: 'onblueberry' }]} />} mode={'onblueberry'} />
    <RadioButton {...args} label={<Label labelTexts={[{ text: 'oninvalid' }]} />} mode={'oninvalid'} />
    <RadioButton {...args} label={<Label labelTexts={[{ text: 'onwhite - disabled' }]} />} mode={'onwhite'} disabled />
    <div style={{ backgroundColor: '#06596C', display: 'block', marginTop: '1rem', padding: '1rem' }}>
      <RadioButton {...args} label={<Label labelTexts={[{ text: 'ondark' }]} />} mode={'ondark'} />
    </div>
  </GridExample>
);

export const Large: ComponentStory<typeof RadioButton> = args => (
  <GridExample>
    <FormGroup legend={'onwhite'} name="radio1" mode={'onwhite'} size={'large'}>
      <RadioButton {...args} label={<Label labelTexts={[{ text: 'onwhite' }]} />} />
      <RadioButton {...args} label={<Label labelTexts={[{ text: 'onwhite' }]} />} />
    </FormGroup>
    <FormGroup legend={'ongrey'} name="radio2" mode={'ongrey'} size={'large'}>
      <RadioButton {...args} label={<Label labelTexts={[{ text: 'ongrey' }]} />} />
      <RadioButton {...args} label={<Label labelTexts={[{ text: 'ongrey' }]} />} />
    </FormGroup>
    <FormGroup legend={'onblueberry'} name="radio3" mode={'onblueberry'} size={'large'}>
      <RadioButton {...args} label={<Label labelTexts={[{ text: 'onblueberry' }]} />} />
      <RadioButton {...args} label={<Label labelTexts={[{ text: 'onblueberry' }]} />} />
    </FormGroup>
    <div style={{ background: getColor('blueberry', 500), padding: '2rem' }}>
      <FormGroup legend={'ondark'} name="radio4" mode={'ondark'} size={'large'}>
        <RadioButton {...args} label={<Label labelTexts={[{ text: 'ondark' }]} />} />
        <RadioButton {...args} label={<Label labelTexts={[{ text: 'ondark' }]} />} />
      </FormGroup>
    </div>

    <RadioButton {...args} label={<Label labelTexts={[{ text: 'oninvalid' }]} />} size={'large'} mode={'oninvalid'} />
    <RadioButton {...args} label={<Label labelTexts={[{ text: 'onwhite - disabled' }]} />} size={'large'} mode={'onwhite'} disabled />
  </GridExample>
);
