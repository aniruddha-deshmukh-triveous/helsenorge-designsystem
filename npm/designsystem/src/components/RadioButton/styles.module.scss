@import '../../scss/_spacers.scss';
@import '../../scss/_breakpoints.scss';
@import '../../scss/_palette.scss';
@import '../../scss/font-settings.scss';

.radio-button-errors {
  font-size: $font-size-sm;
  font-weight: 600;
  color: $cherry500;
  margin-top: 0;
}

.radio-button-wrapper {
  cursor: pointer;
  margin: getSpacer(s) 0;

  &--with-error {
    padding: getSpacer(s) getSpacer(xs) getSpacer(m);
    background-color: $cherry50;
    margin-left: -0.75rem;
    border-left: 0.25rem solid $cherry500;

    @media (min-width: map-get($grid-breakpoints, md)) {
      padding: getSpacer(s) getSpacer(m) getSpacer(l);
    }
  }

  &--with-error#{&}--bigform {
    padding: getSpacer(m) getSpacer(m);
  }

  &--bigform {
    margin: getSpacer(xs) 0 getSpacer(xs) 0;
  }
}

.radio-button-wrapper:first-of-type {
  margin-top: 0;
}

.radio-button-wrapper:last-of-type {
  margin-bottom: 0;
}

.radio-button-label {
  display: flex;
  align-items: center;
  cursor: pointer;
  font-size: $font-size-sm;

  @media (min-width: map-get($grid-breakpoints, md)) {
    font-size: $font-size-md;
  }

  &--on-dark {
    color: $white;
  }

  &--invalid {
    color: $black;
  }

  &--disabled {
    cursor: default;
    color: $neutral400;
  }

  &--bigform {
    padding: getSpacer(s);
    background-color: $neutral50;
  }

  &--bigform#{&}--on-blueberry {
    background-color: $blueberry50;
  }

  &--bigform#{&}--on-dark {
    background-color: $blueberry800;
  }

  &--bigform#{&}--invalid {
    background-color: $white;
  }

  @media (min-width: map-get($grid-breakpoints, md)) {
    &--bigform {
      padding: 1.68rem getSpacer(l);
    }
  }
}

.radio-button {
  -moz-appearance: none;
  -webkit-appearance: none;
  -o-appearance: none;

  cursor: pointer;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 10rem;
  margin: getSpacer(3xs) getSpacer(s) getSpacer(3xs) getSpacer(3xs);
  height: 1.5rem;
  width: 1.5rem;
  min-width: auto !important;
  min-height: auto !important;
  color: $neutral700;
  box-shadow: 0 0 0 getSpacer(3xs);

  &:checked {
    color: $blueberry500;

    &:hover {
      background-color: $blueberry50;
    }
  }

  &:focus {
    box-shadow: 0 0 0 getSpacer(2xs);
    color: $black;
  }

  &:hover {
    box-shadow: 0 0 0 getSpacer(2xs);
    background-color: $neutral100;
  }

  :active > & {
    box-shadow: 0 0 0 getSpacer(2xs);
    color: $black;
  }

  &--on-blueberry {
    color: $blueberry500;

    :hover > & {
      background-color: $blueberry100;
    }
  }

  &--on-dark {
    color: $neutral200;

    &:hover {
      background-color: $inverted-hover;
      box-shadow: 0 0 0 getSpacer(2xs);
    }

    &:checked {
      color: $blueberry200;

      &:hover {
        background-color: $inverted-hover;
      }
    }

    :active > & {
      box-shadow: 0 0 0 getSpacer(2xs);
      color: $white;
    }

    &:focus {
      color: $white;
    }
  }

  &--invalid {
    color: $cherry500;

    &:hover {
      background-color: $cherry100;
    }

    &:checked {
      color: $cherry500;

      &:hover {
        background-color: $cherry100;
      }
    }

    &:active {
      color: $black;
    }

    &:focus {
      color: $black;
    }
  }

  &--disabled {
    cursor: default;

    color: $neutral400;
    &:checked {
      color: $neutral400;

      &:hover {
        box-shadow: 0 0 0 getSpacer(3xs);
        background-color: transparent;
      }
    }
    &:hover {
      box-shadow: 0 0 0 getSpacer(3xs);
      background-color: transparent;
    }

    &:active {
      color: $neutral400;
    }
  }

  // :before er dot i RadioButton
  &:checked::before {
    content: '';
    display: flex;
    background-color: $blueberry500;
    width: 0.75rem;
    height: 0.75rem;
    border-radius: 10rem;
  }

  &--on-dark:checked::before {
    background-color: $blueberry200;
  }

  &--disabled:checked::before {
    box-shadow: 0 0 0 0;
    background-color: $neutral400;
    color: $neutral400;
  }

  &--invalid:checked::before {
    background-color: $cherry500;
  }
}
