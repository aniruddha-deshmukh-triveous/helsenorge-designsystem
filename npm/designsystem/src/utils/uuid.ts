export const uuid = (): string => {
  if (process.env.NODE_ENV === 'test') {
    return 'test';
  }
  return 'bxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx'.replace(/[x]/g, function (c) {
    const r = (Math.random() * 16) | 0,
      v = c == 'x' ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
};

export default uuid;
