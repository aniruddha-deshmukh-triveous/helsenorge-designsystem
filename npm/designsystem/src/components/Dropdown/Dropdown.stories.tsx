import React from 'react';
import { withKnobs, text, select, boolean } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import { withA11y } from '@storybook/addon-a11y';
import Dropdown, { DropdownMode } from './Dropdown';
import Checkbox from '../Checkbox';
import RadioButton from '../RadioButton';

const stories = storiesOf('Dropdown', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withA11y);

stories.add('RadioButton', () => (
  <div style={{ width: '60rem' }}>
    <Dropdown
      label={text('Dropdown label', 'Ta et valg')}
      toggleLabel={text('Toggle label', 'Hva skjer i kroppen?')}
      closeText={text('Close text', 'Lukk')}
      open={boolean('Open', false)}
      mode={select('Mode', DropdownMode, DropdownMode.onwhite)}
      transparent={boolean('Transparent', false)}
      fluid={boolean('Fluid', false)}
      noCloseButton={boolean('No close button', false)}
      disabled={boolean('Disabled', false)}
      onToggle={value => console.log('onToggle', value)}
    >
      <RadioButton label="Valg 1" name="radiobutton" />
      <RadioButton label="Valg 2" name="radiobutton" />
      <RadioButton label="Valg 3" name="radiobutton" />
      <RadioButton label="Valg 4" name="radiobutton" />
      <RadioButton label="Valg 5" name="radiobutton" />
      <RadioButton label="Valg 6" name="radiobutton" />
      <RadioButton label="Valg 7" name="radiobutton" />
      <RadioButton label="Valg 8" name="radiobutton" />
      <RadioButton label="Valg 9" name="radiobutton" />
    </Dropdown>
  </div>
));
stories.add('Checkbox', () => (
  <div style={{ width: '60rem' }}>
    <Dropdown
      label={text('Dropdown label', 'Ta et valg')}
      toggleLabel={text('Toggle label', 'Hva skjer i kroppen?')}
      closeText={text('Close text', 'Lukk')}
      open={boolean('Open', false)}
      mode={select('Mode', DropdownMode, DropdownMode.onwhite)}
      transparent={boolean('Transparent', false)}
      fluid={boolean('Fluid', false)}
      noCloseButton={boolean('No close button', false)}
      disabled={boolean('Disabled', false)}
      onToggle={value => console.log('onToggle', value)}
    >
      <Checkbox label="Valg 1" name="checkbox" />
      <Checkbox label="Valg 2" name="checkbox" />
      <Checkbox label="Valg 3" name="checkbox" />
      <Checkbox label="Valg 4" name="checkbox" />
      <Checkbox label="Valg 5" name="checkbox" />
      <Checkbox label="Valg 6" name="checkbox" />
      <Checkbox label="Valg 7" name="checkbox" />
      <Checkbox label="Valg 8" name="checkbox" />
      <Checkbox label="Valg 9" name="checkbox" />
    </Dropdown>
  </div>
));
stories.add('Custom content', () => (
  <div style={{ width: '60rem' }}>
    <Dropdown
      label={text('Dropdown label', 'Ta et valg')}
      toggleLabel={text('Toggle label', 'Hva skjer i kroppen?')}
      closeText={text('Close text', 'Lukk')}
      open={boolean('Open', false)}
      mode={select('Mode', DropdownMode, DropdownMode.onwhite)}
      transparent={boolean('Transparent', false)}
      fluid={boolean('Fluid', false)}
      noCloseButton={boolean('No close button', false)}
      disabled={boolean('Disabled', false)}
      onToggle={value => console.log('onToggle', value)}
    >
      <div style={{ padding: '1rem' }}>
        {
          'Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen? Hva skjer i kroppen?'
        }
      </div>
    </Dropdown>
  </div>
));
