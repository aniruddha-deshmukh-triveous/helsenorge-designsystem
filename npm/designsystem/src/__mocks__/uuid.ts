import { vi as jest } from 'vitest';

import * as uuidUtils from '../utils/uuid';

const testId = {
  id: 0,
};

/**
 * Returnerer en unik, forutsigbar id hver gang uuid() kalles
 */
jest.spyOn(uuidUtils, 'uuid').mockImplementation(() => {
  testId.id++;
  return `testid-${testId.id}`;
});
