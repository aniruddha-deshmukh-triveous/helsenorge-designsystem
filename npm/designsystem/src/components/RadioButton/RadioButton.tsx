import React from 'react';

import classNames from 'classnames';

import { AnalyticsId, FormMode, FormVariant } from '../../constants';
import { uuid } from '../../utils/uuid';

import radioButtonStyles from './styles.module.scss';

export interface RadioButtonProps
  extends Pick<React.InputHTMLAttributes<HTMLInputElement>, 'name' | 'value' | 'disabled' | 'defaultChecked' | 'required'> {
  /** Adds custom classes to the element. */
  className?: string;
  /** The label text next to the radioButton */
  label: string;
  /** input id of the radioButton */
  inputId?: string;
  /** Changes the visuals of the radioButton */
  mode?: keyof typeof FormMode;
  /** Changes the visuals of the radioButton */
  variant?: keyof typeof FormVariant;
  /** Activates Error style for the radioButton - This is can be true while errorText is empty, when in a FormGroup */
  error?: boolean;
  /** Error text to show above the component */
  errorText?: string;
  /** Sets the data-testid attribute. */
  testId?: string;
}

export const RadioButton = React.forwardRef((props: RadioButtonProps, ref: React.Ref<HTMLInputElement>) => {
  const {
    className,
    defaultChecked = false,
    disabled,
    label,
    inputId = uuid(),
    mode,
    name = inputId,
    variant,
    errorText,
    error = !!errorText,
    value = label,
    testId,
    required,
    ...rest
  } = props;
  const invalid = error || mode === FormMode.oninvalid;
  const onDark = mode === FormMode.ondark;
  const onBlueberry = mode === FormMode.onblueberry;
  const bigform = variant === FormVariant.bigform;

  const radioButtonWrapperClasses = classNames(radioButtonStyles['radio-button-wrapper'], {
    [radioButtonStyles['radio-button-wrapper--with-error']]: errorText,
    [radioButtonStyles['radio-button-wrapper--bigform']]: bigform,
  });
  const radioButtonLabelClasses = classNames(radioButtonStyles['radio-button-label'], {
    [radioButtonStyles['radio-button-label--disabled']]: disabled,
    [radioButtonStyles['radio-button-label--on-dark']]: onDark,
    [radioButtonStyles['radio-button-label--on-blueberry']]: onBlueberry,
    [radioButtonStyles['radio-button-label--invalid']]: invalid,
    [radioButtonStyles['radio-button-label--bigform']]: bigform,
  });
  const radioButtonClasses = classNames(
    radioButtonStyles['radio-button'],
    {
      [radioButtonStyles['radio-button--on-dark']]: onDark,
      [radioButtonStyles['radio-button--disabled']]: disabled,
      [radioButtonStyles['radio-button--on-blueberry']]: onBlueberry,
      [radioButtonStyles['radio-button--invalid']]: invalid,
    },
    className
  );
  const errorStyles = classNames(radioButtonStyles['radio-button-errors'], {
    [radioButtonStyles['radio-button-errors--bigform']]: bigform,
  });

  return (
    <div data-testid={testId} data-analyticsid={AnalyticsId.RadioButton} className={radioButtonWrapperClasses}>
      {errorText && <p className={errorStyles}>{errorText}</p>}
      <label htmlFor={inputId} className={radioButtonLabelClasses}>
        <input
          id={inputId}
          name={name}
          className={radioButtonClasses}
          type="radio"
          disabled={disabled}
          value={value}
          ref={ref}
          aria-invalid={error}
          defaultChecked={defaultChecked}
          required={required}
          {...rest}
        />
        {label}
      </label>
    </div>
  );
});

export default RadioButton;
