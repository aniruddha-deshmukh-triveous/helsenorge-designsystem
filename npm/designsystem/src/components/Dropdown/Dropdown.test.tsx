import React from 'react';
import { render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Dropdown from './';
import RadioButton from '../RadioButton';

import '../../__mocks__/uuid';

describe('Gitt at Dropdown skal vises vanlig', (): void => {
  describe('Når Dropdownen har vanlig innhold', (): void => {
    test('Så kan man klikke og se innholdet', (): void => {
      render(
        <Dropdown label="Ta et valg" toggleLabel="Knapp">
          <h2>Innhold i Dropdown</h2>
        </Dropdown>
      );

      const button = screen.getByRole('button', { name: 'Knapp' });

      userEvent.click(button);

      const options = screen.getByLabelText('Ta et valg');
      expect(options).toHaveFocus();

      const contents = within(options).getByRole('heading', { name: 'Innhold i Dropdown' });
      expect(contents).toBeVisible();
    });

    test('Så ser Dropdown ut slik den skal', (): void => {
      const { container } = render(
        <Dropdown label="Ta et valg" toggleLabel="Knapp">
          <h2>Innhold i Dropdown</h2>
        </Dropdown>
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('Når Dropdownen har lukkeknapp', (): void => {
    test('Så kan man lukke dropdownen med knappen', (): void => {
      render(
        <Dropdown label="Ta et valg" toggleLabel="Knapp" closeText="Lukk">
          <h2>Innhold i Dropdown</h2>
        </Dropdown>
      );

      const button = screen.getByRole('button', { name: 'Knapp' });
      expect(button).toHaveAttribute('aria-expanded', 'false');
      userEvent.click(button);
      expect(button).toHaveAttribute('aria-expanded', 'true');
      const close = screen.getByRole('button', { name: 'Lukk' });
      userEvent.click(close);
      expect(close).toHaveAttribute('aria-expanded', 'false');
    });
  });

  describe('Når Dropdownen har open prop', (): void => {
    test('Så er den åpen', (): void => {
      render(
        <Dropdown label="Ta et valg" toggleLabel="Knapp" open>
          <h2>Innhold i Dropdown</h2>
        </Dropdown>
      );

      const button = screen.getByRole('button', { name: 'Knapp' });
      expect(button).toHaveAttribute('aria-expanded', 'true');
    });
  });

  describe('Når Dropdownen har disabled prop', (): void => {
    test('Så er den disabled', (): void => {
      render(
        <Dropdown label="Ta et valg" toggleLabel="Knapp" disabled>
          <h2>Innhold i Dropdown</h2>
        </Dropdown>
      );

      const button = screen.getByRole('button', { name: 'Knapp' });
      expect(button).toBeDisabled();
    });
  });

  describe('Når Dropdownen har noCloseButton prop', (): void => {
    test('Så er det ingen lukkeknapp', (): void => {
      render(
        <Dropdown label="Ta et valg" toggleLabel="Knapp" closeText="Lukk" noCloseButton>
          <h2>Innhold i Dropdown</h2>
        </Dropdown>
      );

      const button = screen.getByRole('button', { name: 'Knapp' });
      userEvent.click(button);

      const close = screen.queryByRole('button', { name: 'Lukk' });
      expect(close).not.toBeInTheDocument();
    });
  });

  describe('Når man klikker utenfor dropdownen mens den er åpen', (): void => {
    test('Så lukkes Dropdownen', (): void => {
      render(
        <>
          <button>Knapp utenfor dropdown</button>
          <Dropdown label="Ta et valg" toggleLabel="Knapp" open>
            <h2>Innhold i Dropdown</h2>
          </Dropdown>
        </>
      );

      const button = screen.getByRole('button', { name: 'Knapp' });
      expect(button).toHaveAttribute('aria-expanded', 'true');

      const outsideButton = screen.getByRole('button', { name: 'Knapp utenfor dropdown' });
      userEvent.click(outsideButton);

      expect(button).toHaveAttribute('aria-expanded', 'false');
    });
  });

  describe('Når man klikker på Escape mens Dropdown er åpen', (): void => {
    test('Så lukkes Dropdownen', (): void => {
      render(
        <Dropdown label="Ta et valg" toggleLabel="Knapp">
          <h2>Innhold i Dropdown</h2>
        </Dropdown>
      );
      const button = screen.getByRole('button', { name: 'Knapp' });
      userEvent.click(button);
      userEvent.keyboard('{esc}');

      expect(button).toHaveAttribute('aria-expanded', 'false');
    });
  });

  describe('Når man klikker på Enter mens Dropdown er lukket', (): void => {
    test('Så åpnes Dropdownen', (): void => {
      render(
        <Dropdown label="Ta et valg" toggleLabel="Knapp">
          <h2>Innhold i Dropdown</h2>
        </Dropdown>
      );
      const button = screen.getByRole('button', { name: 'Knapp' });
      // Åpne og lukke dropdownen slik at fokus er på knappen før vi klikker på tastaturet
      userEvent.click(button);
      userEvent.click(button);
      userEvent.keyboard('{enter}');

      expect(button).toHaveAttribute('aria-expanded', 'true');
    });
  });

  describe('Når man klikker på Home mens Dropdown er åpen', (): void => {
    test('Så flyttes fokus til første radioknapp, og man kan bruke tastaturet for å gå nedover, men ikke oppover', (): void => {
      render(
        <Dropdown label="Ta et valg" toggleLabel="Knapp">
          <RadioButton label="Førstevalg" inputId="radio-1" value="radio-1" />
          <RadioButton label="Andrevalg" inputId="radio-2" value="radio-2" />
          <RadioButton label="Tredjevalg" inputId="radio-3" value="radio-3" />
        </Dropdown>
      );
      const button = screen.getByRole('button', { name: 'Knapp' });
      userEvent.click(button);
      userEvent.keyboard('{home}');

      const radio1 = screen.getByLabelText('Førstevalg');
      expect(radio1).toHaveFocus();

      userEvent.keyboard('{arrowup}');
      expect(radio1).toHaveFocus();

      userEvent.keyboard('{arrowdown}');
      const radio2 = screen.getByLabelText('Andrevalg');
      expect(radio2).toHaveFocus();
    });
  });

  describe('Når man klikker på End mens Dropdown er åpen', (): void => {
    test('Så flyttes fokus til sist radioknapp, og man kan bruke tastaturet for å gå oppover, men ikke nedover', (): void => {
      render(
        <Dropdown label="Ta et valg" toggleLabel="Knapp">
          <RadioButton label="Førstevalg" inputId="radio-1" value="radio-1" />
          <RadioButton label="Andrevalg" inputId="radio-2" value="radio-2" />
          <RadioButton label="Tredjevalg" inputId="radio-3" value="radio-3" />
        </Dropdown>
      );
      const button = screen.getByRole('button', { name: 'Knapp' });
      userEvent.click(button);
      userEvent.keyboard('{end}');

      const radio3 = screen.getByLabelText('Tredjevalg');
      expect(radio3).toHaveFocus();

      userEvent.keyboard('{arrowdown}');
      expect(radio3).toHaveFocus();

      userEvent.keyboard('{arrowup}');

      const radio2 = screen.getByLabelText('Andrevalg');
      expect(radio2).toHaveFocus();
    });
  });
});
