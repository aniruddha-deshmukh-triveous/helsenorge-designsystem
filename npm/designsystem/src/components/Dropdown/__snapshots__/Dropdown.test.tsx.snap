// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at Dropdown skal vises vanlig Når Dropdownen har vanlig innhold Så ser Dropdown ut slik den skal 1`] = `
<div>
  <div
    class="dropdown"
  >
    <span
      class="dropdown__label"
      id="testid-17"
    >
      Ta et valg
    </span>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="testid-18"
      class="dropdown__toggle dropdown__toggle--on-white"
      data-analyticsid="dropdown"
      type="button"
    >
      <span
        class="dropdown__toggle__label"
        id="testid-18"
      >
        Knapp
      </span>
      <svg
        aria-hidden="true"
        class="hnds-style-icon dropdown__icon"
        data-analyticsid="icon"
        fill="#126F87"
        focusable="false"
        height="38"
        role="presentation"
        style="min-width: 38px; min-height: 38px;"
        viewBox="0 0 48 48"
        width="38"
      >
        <path
          d="M24 14.18v8.556h8.558v2.527H24v8.558h-2.526l-.001-8.558h-8.557v-2.526h8.557v-8.558H24z"
        />
      </svg>
    </button>
    <div
      class="dropdown__content"
    >
      <ul
        aria-labelledby="testid-17"
        class="dropdown__options"
        role="listbox"
        tabindex="-1"
      >
        <li
          aria-selected="false"
          class="dropdown__input"
          id="testid-19-0"
          role="option"
        >
          <h2>
            Innhold i Dropdown
          </h2>
        </li>
      </ul>
      <div
        class="dropdown__close"
      >
        <button
          aria-expanded="false"
          class="button-wrapper button-wrapper--fluid"
          data-analyticsid="button"
          type="button"
        >
          <div
            class="button button--normal"
          >
            <div
              class="button__text"
            >
              <span>
                Lukk
              </span>
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
